src/libraries/System.Linq/tests/CountByTests.cs

```diff
@@ -32,29 +32,29 @@ public void CountBy_SourceThrowsOnGetEnumerator()
         {
             IEnumerable<int> source = new ThrowsOnGetEnumerator();
 
-            var enumerator = source.CountBy(x => x).GetEnumerator();
+            var countBy = source.CountBy(x => x);
 
-            Assert.Throws<InvalidOperationException>(() => enumerator.MoveNext());
+            Assert.Throws<InvalidOperationException>(() => countBy.GetEnumerator().MoveNext());
```

Comment:
I'm guessing you meant to assert this?
```suggestion
            Assert.Throws<InvalidOperationException>(() => countBy.GetEnumerator());
```
I don't think this is an acceptable behavioural change. Generally speaking `GetEnumerator()` calls don't do any meaningful work other than instantiate the enumerator; everything else should be reserved for the first `MoveNext()` call, including any nested `GetEnumerator()` calls.

Reply:
I think `countBy.GetEnumerator().MoveNext()` is independent of how the implementation is done; If the implementation changes, the test will not have to change.

Reply:
The `MoveNext()` part is irrelevant in this case, it is `countBy.GetEnumerator()` that is throwing in this case when it previously wasn't. `GetEnumerator` shouldn't be doing any work or throwing exceptions, however with this change it is clear that `GetEnumerator` is doing _all_ the work (since it is this method that is enumerating the entire source enumerable to build the dictionary).

Reply:
As far as I know, in the `IQueryable` interface it is the `GetEnumerator` method that does the work of initializing the query, creating an SQL query, sending the SQL query to the server, passing the query parameters, creating a cursor to obtain the result using the MoveNext method, etc. It can also throw exceptions.

Therefore, I think that the `IEnumerable.GetEnumerator` method can work even before the first call to MoveNext, including throwing exceptions.

I think that the work can be done both as in the GetEnumerator method and as in the first call to MoveNext. These are already implementation details, so I wrote the tests so that they do not depend on the subtleties of the implementation.

Reply:
This is not an apples-to-apples comparison. `IQueryable` uses a very different execution model, especially when building SQL queries. But even if we look at how the `IEnumerable` implementation of built-in query provider works the same invariant holds:

```C#
IQueryable<int> queryable = new ThrowingGetEnumerator().AsQueryable()
    .Select(x => x + 1)
    .Where(x => x > 0);

IEnumerator<int> enumerator = queryable.GetEnumerator(); // No exception
enumerator.MoveNext(); // Exception

public class ThrowingGetEnumerator : IEnumerable<int>
{
    public IEnumerator<int> GetEnumerator() => throw new NotSupportedException();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}
```

> These are already implementation details, so I wrote the tests so that they do not depend on the subtleties of the implementation.

I strongly disagree that that it's an implementation detail. Changing the behavior can break user code that depends on this method not enumerating the source or throwing exceptions; it might not be important if you just foreach over the result, but it is in code that explicitly handles enumerators in bespoke ways.

Reply:
Add `GroupBy(x => x)` to query!
`GroupBy` is analog to how CountBy works and `queryable.GetEnumerator()` will throw an exception.

See https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0AwgPYA2FApgMbACWZUAzgNw76EAsHGXxADIMoARz45MATgAUAIgCKEamACecgJR8AbgEMweUcrW6ARjTwBePFGoB3PABUAFmDJ2RAcwDi1YAFEoCABbFV1gMjAZDSIAQRYlFVUzGmj0PAy8IgBlahp6GQAPKwA+PGKAajxMDXTMogB1ZxVqItLyvDKkWrRMrO83CAAHACFVNssywq10SVk5XwCg0LBwyM0dfTxqZbCIg2sjJJTqIkXAkL3I6IkMaXkAWTJtagA5akLgDfQdy9X9ohPF7vT43WZ3ebZCC0WjUFgsb5oHAAZgICCcrncXnOu3+kTwIAImGRAB4RMASugAN51DJwVEASQuKzWYDJUApeBxf1Z0XawExDlsDleZGAUKGQ0iwGoABN/IVYUNGMwwb1Mkzcayicizn5mVcohp2tyWfs1QBfIA===

Reply:
I would argue that this indicates an issue with the `GroupBy` implementation. `GetEnumerator()` should not be incurring side-effects in principle.

Reply:
I guess `GetEnumerator` can start to execute work before the first call `MoveNext` and it depends on implementation.